### ISSUE TITLE
Refactor PositionIterator

### ISSUE THREAD

**Issue by pjasiun** [ckeditor/ckeditor5-core#65]
`PositionIterator` keeps only the `Position` so it calculates a node and offset every time the `next()` method is called and recalculates a new position based on the new node and offset.

This could be improved by caching node and offset, but it will not work if the document will change during iteration.

On the other hand, if document changes, the position makes no sense, so we could assume that document is immutable during iterating.

Note that then, iterator should work differently for `SmartPosition` (see #64). There changing document during iteration makes sense.

**Comment by pjasiun** [comment]
Or there could be two iterators: `Static` and `Dynamic`, because caching may improve also iterating `SmartPosition` as long as document does not change during iteration.

**Comment by scofalik** [comment]
In my implementation of simple Tree Model for OT purposes, my `Node` class had properties `.prevNode` and `.nextNode` pointing to, well, previous and next siblings of that node in the structure. Those were updated whenever a node has been added or removed from the tree (this is fairly easy operation done in a static amount of time). If you had these, you could cache the last visited node and use it's `nextNode` and `prevNode` properties. Whenever `nextNode` or `prevNode` is `null`, you know that you reached the last sibling and you have to "move up". I think this approach is less affected by changes on the document during iteration. You could easily remove some nodes during iteration and still have a valid iterator state. Of course if you move the last visited node *somewhere* strange things may happen :).

By the way, having `nextSibling` and `previousSibling` as real, set properties that are not re-evaulated each time they are called will bring some optimization anyway, so I propose to change them. I am sure you don't need code snippet for that, but if you want to have a quick look how it could look, you may check out: https://github.com/ckeditor/ckeditor5-design/blob/op-trans/ot_ranges/ot.js#L44-L63

**Comment by pjasiun** [comment]
Now we can not store `nextSibling` and `previousSibling` in each node, because it cost too much memory. Now we store only `character` and `parent` per each `Character` node. With `nextSibling` and `previousSibling` it would be more-less twice that much. 

But I hope we will introduce text nodes compression (#62) soon. Then it would be a very nice solution.

**Comment by Reinmar** [comment]
I'm not sure how and when iterators will be used as we don't have any use cases implemented yet. However, in CKE4 we always had the same issue that our ranges or walker may easily get desynced. We got used to that and learned to deal with the situation, so I would say that we can live with such iterator in CKE5 too.

Second thing – natively you have two iterators as well – `TreeWalker` and `NodeIterator`:

> NodeIterators and TreeWalkers are dynamic - the logical view changes to reflect changes made to the underlying document. However, they differ in how they respond to those changes. NodeIterators, which present the nodes sequentially, attempt to maintain their location relative to a position in that sequence when the sequence's contents change. TreeWalkers, which present the nodes as a filtered tree, maintain their location relative to their current node and remain attached to that node if it is moved to a new context. We will discuss these behaviors in greater detail below.

(Source: http://www.w3.org/TR/DOM-Level-2-Traversal-Range/traversal.html)

Note how their behaviour would differ in case of document changes. In both cases you can apply such changes which will invalidate the iterator. So no matter how `SmartPosition` is going to work, you may have situations which will be hard to resolve in a safe way (current subtree was moved from one place to another, the whole subtree was removed, etc). So IMO **you expect the tree to be immutable while iterating over it**. It's just like with loops – you don't want to modify the array because you need to control what to do with your index which raises complexity of the algorithm. In the tree gets even more complicated.

So to sum up – I don't like how the `PositionIterator` works right now. Basing it on `path` created very fragile and slow construction. I think that we can safely assume that the document is immutable (at least locally) and base the iterator on container and offset. Clarifying this in the documentation will be enough.

**Comment by Reinmar** [comment]
Three more things:

* Do we have any clear use cases for position iterator?
* Why "position iterator" and not something more usual like "tree walker"?
* I don't understand the `Position` class as well (see #83).

**Comment by pjasiun** [comment]
> Do we have any clear use cases for position iterator?

You can find a usage here: https://github.com/ckeditor/ckeditor5-core/blob/master/src/document/operation/changeoperation.js#L105-L119

> Why "position iterator" and not something more usual like "tree walker"?

Because it is a JavaScript  iterator (with the `next` method). But "tree walker" is fine for me too.

> I don't understand the Position class as well (see #83).

I replied in #83.

> So no matter how SmartPosition is going to work, you may have situations which will be hard to resolve in a safe way.

I think that iterator for `SmartPosition` may work on the document which is mutable. But you can not think about it as a walker which return all nodes in the range no matter what, but as a structure which will return the next position as long it exists and you are not out of the boundary range. I also agree that we have no real use case for such structure so the dynamic iterator should be implemented only if needed.

Now I think we agreed that document is immutable during the iteration, so we can introduce optimization for the iterator.

**Comment by Reinmar** [comment]
> > Why "position iterator" and not something more usual like "tree walker"?
>
> Because it is a JavaScript iterator (with the next method). But "tree walker" is fine for me too.

OK, makes sense that it's called an iterator, but it's not *position* iterator. It's a range iterator. Position is just one point, while to iterate you need two points (hence, range).

**Comment by Reinmar** [comment]
> I also agree that we have no real use case for such structure so the dynamic iterator should be implemented only if needed.

Agree. Once we have a use case and if we find it feasible, we can have it.

> Now I think we agreed that document is immutable during the iteration, so we can introduce optimization for the iterator.

:+1:

**Comment by Reinmar** [comment]
> You can find a usage here: https://github.com/ckeditor/ckeditor5-core/blob/master/src/document/operation/changeoperation.js#L105-L119

Again, super clear that the class should be called a `RangeIterator` :D. Plus, we should find a better name for the "item" than `value`. Isn't it "position"?

**Comment by pjasiun** [comment]
> OK, makes sense that it's called an iterator, but it's not position iterator. It's a range iterator. Position is just one point, while to iterate you need two points (hence, range).

No, you do not. You can create `PositionIterator` from the single position, which because the starting position of the iterator. There will be no boundary then so it will iterate till the end of the tree.

**Comment by Reinmar** [comment]
> > OK, makes sense that it's called an iterator, but it's not position iterator. It's a range iterator. Position is just one point, while to iterate you need two points (hence, range).
>
> No, you do not. You can create PositionIterator from the single position, which because the starting position of the iterator. There will be no boundary then so it will iterate till the end of the tree.

But these are still two points. One is inside the document and one is at the end of it. Position is a point and semantically it has no length, so it cannot be "iterated over".

**Comment by pjasiun** [comment]
> But these are still two points.

No, there are always 3 positions: current position of the iterator, begging of the boundary and the end of the boundary. You can iterate back using the `previous` method. This class was created not only to iterate whole range at once, but also to move the position inside the document in the smart way. Anyway I was also thinking about the name `RangeIterator`, it was the actually the previous name of it.

**Comment by pjasiun** [comment]
> Plus, we should find a better name for the "item" than value. Isn't it "position"?

And again no ;) When you call method `next` the iterator set its position to the next position, but what it returns is the value it consumed with the type. So the `value` have two properties, one is `type` which can be `ELEMENT_ENTER`, `ELEMENT_LEAVE` or `CHARACTER`, and the second is `node`. I would love to find a better name for such pair, but I got nothing.

**Comment by pjasiun** [comment]
BTW: talking about that class I realized that there is a bug in the only usage of it, so I reported a ticket: https://github.com/ckeditor/ckeditor5-core/issues/84.

**Comment by Reinmar** [comment]
> No, there are always 3 positions:

I'm sorry PJ, but your explanation makes it only more apparent that it should not be "iterator" and it has little to do with a single position. In such case, I'm all for `TreeWalker` (to underline that you can do more than just iterator) or `RangeIterator` to KISS.

**Comment by pjasiun** [comment]
> I'm sorry PJ, but your explanation makes it only more apparent that it should not be "iterator" and it has little to do with a single position. In such case, I'm all for TreeWalker (to underline that you can do more than just iterator) or RangeIterator to KISS.

Yep, it makes sens to rename it. I like the name `TreeWalker`, because maybe we will add more walking methods in the future (moving to the begging/end of the element, or parent?).

**Comment by scofalik** [comment]
This issue will also include renaming `PositionIterator`. Probably some refactor will be also introduced.

**Comment by scofalik** [comment]
`PositionIterator` got renamed to to `TreeWalker`. We should get used to the new name.

Node caching certainly makes `TreeWalker` faster. The decision was whether we cache most recently visited node or the parent of that node. Most recently node seems like a better fit but it turns out that 1) the implementation would be a bit more complicated and 2) that getting next node to iterate over is actually faster when you have parent (by a very-small margin, but `next/previousSibling` uses `node.parent` anyway). So, last visited parent is now cached.

Another optimization is in creating next/previous position of `TreeWalker`. Before we were creating positions basing on last visited node. The code was simpler and easier to read, but by creating `Position` this way, we have to calculate `node.path`. Now, new position is calculated by hand.

I think that for now `TreeWalker` is okay. It's hard to find a place for real improvement, especially performance-wise. I included described changes in PR #139 .

**Comment by pjasiun** [comment]
I checked changes introduced in https://github.com/ckeditor/ckeditor5-core/pull/139 and they are fine. I pushed small docs change directly to the master branch and the ticket can be closed.
